File: 백준/Silver/9012. 괄호/괄호.cc
#include <stdio.h>
#include <string.h>

int main() {
    int n;
    scanf("%d", &n);

    char p[1001];

    for (int i = 0; i < n; i++) {
        scanf("%s", p);  

        int cnt = 0;     
        int isValid = 1; 


        for (int j = 0; p[j] != '\0'; j++) {
            if (p[j] == '(') {
                cnt++;   
            } else {      
                if (cnt == 0) {
                    isValid = 0;
                    break;
                }
                cnt--;
            }
        }

        if (cnt != 0) {
            isValid = 0;
        }

        if (isValid)
            printf("YES\n");
        else
            printf("NO\n");
    }

    return 0;
}

File: 백준/Silver/9012. 괄호/괄호.swift
let n = Int(readLine()!)!
var ps: [String] = []

for i in 1...n {
    let s = readLine()!
    ps.append(s)
}

for p in ps {
    var cnt: Int = 0
    var isValid = true
    
    for ch in p {
        if ch == "(" {
            cnt += 1
        } else {
            if cnt == 0 {
                isValid = false
                break
            }
            cnt -= 1
        }
    }
    
    if cnt != 0 {
        isValid = false
    }
    
    print(isValid ? "YES" : "NO")
}